### PR TITLE
SE-3057: Add Private Markets and Real Estate notebook

### PR DESCRIPTION
**Pull request by dgbarclay**
# Pull Request Checklist

- [x] Changes follow the [style guide](https://github.com/finbourne/sample-notebooks/blob/master/docs/FINBOURNE%20notebook%20style%20guide.ipynb)
- [x] Tests pass
- [x] Raised the PR against the `develop` branch

# Description of the PR

This notebook will show the creation of a private markets and real estate environment, suitable for modelling cash flows and portfolio securitization, among other things.

### CODE REVIEW

File: examples/use-cases/private-markets/Private Markets and Real Estate.ipynb [hunk 1]
```diff
@@ -0,0 +1,5070 @@
+{
```
Comment (cscialino) [inline examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>"This notebook will show the creation of a private markets and real estate environment," --&gt; Sounds a bit techy, I'd change it to "private markets and real estate book,.."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (cscialino) [inline examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    try:
</pre><p>I feel like this code block repeats itself 3x here -- could you wrap it into a function, or run through a loop with the property args in something like a dict? </p><p> </p><p>e.g:</p><pre class="ql-syntax" spellcheck="false">property_args = {
 "expense_type: "Expense Type",
  ...   
}

for k,v in property_args.items():
  create_property(k,v)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (cscialino) [inline examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    portfolio_codes = ["privateEquityFund", "privateDebtFund"]
</pre><p>Same story here as above, if we could consolidate the portfolio creation code blocks to a single cell. Either reading from something like a dict that sets the arguments, or potentially using a file if that seems better given there might be a bit more data to pass through here.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (cscialino) [inline examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>Diagram looks very nice, well done. If not too tricky, might add a legend or something explaining that what's in blue is a portfolio, red a security pointing to it, etc. Essentially a little explainer for the color coding.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (cscialino) [inline examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    # Add our real estate fund
</pre><p>Is this block needed? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (cscialino) [inline examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>I would add a little markdown explainer around transaction config, with a link to a KB article around this as it's a very LUSID specific topic that new readers may struggle to grasp.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (cscialino) [inline examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    gp_calls = capital_calls.iloc[:8]
</pre><p>Feels a little strange to process the file in slices, is it necessary?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (cscialino) [inline examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    transaction_mapping = {
</pre><p>I feel like this mapping has come up a few times in the above. Consider defining it once, and if I'm not mistaken the properties are the only bit that is changes. If so you could try just passing the properties in each segment in place.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

Comment (cscialino) [inline examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    def create_equity(
</pre><p>I wonder if this should be renamed to <code>create_simple_instrument()</code> rather than the equity reference. I would probably remove some of the equity references, not applicable to all instruments. I believe the asset class parameter is required for this instrument definition, may be more suitable to use 'Unknown' for now, but we should also feedback to product that we may need something more appropriate to these asset classes. Something like 'Alternatives' is the first thing to come to mind here. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='38'/>

Comment (cscialino) [inline examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #16.            simple_instrument_type="cmbs", #might not work
</pre><p>I would capitalize "CMBS" and also remove the code comment here. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='47'/>

Comment (cscialino) [inline examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.    transactions.loc[3,'holding_class'] = 'Sydney Bondi Junction Westfield'
</pre><p>Is this block necessary or could we not just have the file ready to go in the correct format from the get go?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='57'/>

Comment (cscialino) [inline examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    real_estate = transactions[3:5]
</pre><p>Necessary to process trades in df slices, or could we use a single call?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='60'/>

Comment (cscialino) [inline examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    transaction_mapping = {
</pre><p>Anything new in this transaction mapping when compared to the earlier ones?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='61'/>

Comment (cscialino) [inline examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #8.                    key="Equity.ClientInternal.*",
</pre><p>Can we add code comments to each section of the recipe, describing what it does? e.g. explain each market data rule, model selection and result data rules. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='68'/>

Comment (cscialino) [inline examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    non_lookthrough_config_recipe = lm.ConfigurationRecipe(
</pre><p>Same as above recipe, add code comments where relevant.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='69'/>

Comment (cscialino) [inline examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>real estate properties --&gt; change to real estate assets/instruments. Wording is fine, but simply clashes with our definition of "properties", so we should try to avoid confusion.</p><p> </p><p>Would rephrase also "However, we care able to store cash flows against any instrument..." and then add a link to a KB article documenting SRS.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='77'/>

Comment (cscialino) [inline examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #5.    london_cf = cash_flow_df[cash_flow_df.client_id == 'LON343239']
</pre><p>Do we need to store these separately? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='78'/>

Comment (cscialino) [inline examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>Would add a little blurb explaining that LUSID valuation engine doing the work here, i.e. projecting the cash flows given it can do so once the economic definition is set. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='80'/>

Comment (cscialino) [inline examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #21.    for val in response.values:
</pre><p>We also have this for the cash ladder in other repos, wonder if it could be moved to a utility folder where it could be re-used here. Other option would be to raise a jira and include in LPT -- see if you can incorporate this here, minimizing the data wrangling you need to do in the story bits of the notebook. </p><p> </p><p>If not, we should raise the jira to move over to something like LPT.</p><p> </p><p><a href="https://github.com/finbourne/sample-notebooks/blob/master/examples/use-cases/instruments/utilities/formatting_tools.py#L83" target="_blank">sample-notebooks/examples/use-cases/instruments/utilities/formatting_tools.py at master · finbourne/sample-notebooks · GitHub</a></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/664/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='83'/>

Comment (csf90-uk) [reply examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>Done.</p>

Comment (csf90-uk) [reply examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>Reworked.</p>

Comment (csf90-uk) [reply examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>Added to markdown cell above.</p>

Comment (csf90-uk) [reply examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>Reworked to be pandaframe filter.</p>

Comment (csf90-uk) [reply examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>Defined as a dictionary and then update only the keys that change throughout.</p>

Comment (csf90-uk) [reply examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>Reworked as per feedback.</p>

Comment (csf90-uk) [reply examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>Done.</p>

Comment (csf90-uk) [reply examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>Updated file with column.</p>

Comment (csf90-uk) [reply examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>Updated to use DF filtering.</p>

Comment (csf90-uk) [reply examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>Reworked.</p>

Comment (csf90-uk) [reply examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>Added comments.</p>

Comment (csf90-uk) [reply examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>Added comments.</p>

Comment (csf90-uk) [reply examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>Updated with suggestions.</p>

Comment (csf90-uk) [reply examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>Reworked although not 100% certain what you meant by the comment.</p>

Comment (csf90-uk) [reply examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>Added extra detail to markdown.</p>

Comment (csf90-uk) [reply examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>I will raise a Jira. Utilities in instruments folder uses deprecated dataframe method (append rather than concat). I reworked it but then the output is quite different from this code so agreed with SLM this existing code is fine.</p>

Comment (csf90-uk) [reply examples/use-cases/private-markets/Private Markets and Real Estate.ipynb:1]: <p>Reworded.</p>

**Review by slemasne — APPROVED** [review]
